library/Fission/CLI/IPFS/Connect.hs

```diff
@@ -28,25 +30,34 @@ swarmConnectWithRetry ::
   , MonadLocalIPFS m
   , MonadWebClient m
   )
-  => IPFS.Peer
+  => NonEmpty IPFS.Peer
   -> Int
   -> m (Either SomeException ())
-swarmConnectWithRetry _peer (-1) =
+swarmConnectWithRetry _peers (-1) =
   return . Left $ toException UnableToConnect
 
-swarmConnectWithRetry peer tries =
-  IPFS.Peer.connect peer >>= \case
-    Right _ ->
+swarmConnectWithRetry peers tries = do
+  attempts <- connectToPeers $ NonEmpty.toList peers
+  if any isRight attempts
+    then 
       return ok
 
-    Left _err ->
+    else 
       Peers.getPeers >>= \case
         Left _ ->
           return . Left $ toException UnableToConnect
 
-        Right peers -> do
+        Right retryPeers -> do
           UTF8.putText "🛰 Unable to connect to the Fission IPFS peer, trying again...\n"
-          swarmConnectWithRetry (head peers) (tries - 1)
+          swarmConnectWithRetry retryPeers (tries - 1)
+
+connectToPeers ::
+  ( MonadUnliftIO  m
+  , MonadLocalIPFS m
+  )
+  => [IPFS.Peer]
+  -> m [Either IPFS.Peer.Error ()]
+connectToPeers peers = sequence $ IPFS.Peer.connect <$> peers
```

Comment:
No retry?

Reply:
the retry is in `swarmConnectWithRetry` which calls this function

Reply:
```haskell
connectToPeers peers = sequence $ connector <$> peers
  where
    connector :: IPFS.Peer -> m ()
    connector peer =  IPFS.Peer.connectWithRetry peer 1
 
-- Or with a monadic fold:

connectToPeers peers = foldM connector (Right ()) peers
  where
    connector :: () -> IPFS.Peer -> m ()
    connector _ peer =  IPFS.Peer.connectWithRetry peer 1
```

We can also attempt these in parallel:

```haskell
sequence $ parMap IPFS.Peer.connect peers
```

Reply:
(mainly unsure how the parMap will interact with go-ipfs)

Reply:
> the retry is in swarmConnectWithRetry which calls this function

Ah, gotcha 👍

Reply:
I'm confused why we need to retry in this function? It's only called in `swarmConnectWithRetry` which handles the retry logic. I could make it a helper function on `swarmConnectWithRetry` to make it more clear?

Reply:
`parMap` makes sense to me though as long as it works with go-ipfs (which I imagine it should)

Reply:
Well, specifically will the daemon try in parallel since there seems to be a lot of singl-threadedness causing headaches. We can probably test this by opening a couple tabs, and asking for peers that don’t exist, and seeing how long between failures (close means actually parallel, spread out means one timed out then the next)

Reply:
Though really adding that one function and testing in the CLI is also pretty lightweight :P

Reply:
just tried it out. it seems that they run in parallel

Reply:
getting compiler errors with `parMap` that I'm not sure how to manage:
```haskell
  attempts <- sequence $ parMap IPFS.Peer.connect (NonEmpty.toList peers)
```
```
    • Couldn't match type ‘Either Network.IPFS.Peer.Error.Error ()’
                     with ‘IPFS.Peer’
      Expected type: Control.Parallel.Strategies.Strategy
                       (Either Network.IPFS.Peer.Error.Error ())
        Actual type: IPFS.Peer
                     -> Control.Parallel.Strategies.Eval
                          (Either Network.IPFS.Peer.Error.Error ())
```
will continue to mess around :thinking: